Django Drip Campaigns blog (#167)
rootstrap/blog: content/Contributing to open source: An overview of the django-drip-campaings project - Bruno Michetti.md

## Contributing to open source: An overview of the django-drip-campaigns project

At Rootstrap we work with many initiatives, one of them is the Open Source one. The main goal in this case is to analyze, maintain and develop open source software (OSS). In this blog I'm going to talk about a particular project named Django Drip Campaigns and how was the process of publishing it as a [pypi](https://pypi.org/) package.  
Pypi (that stands for Python Package Index) is a repository of software for the [Python](https://www.python.org/) programming language that helps us to find and install software developed by the Python community. One of our contributions to the OSS community is to publish and maintain the named project and I want to share that experience in this blog.

### The Django Drip Campaigns Project

Django Drip Campaigns Project is an open source app to create drip campaigns for email using [Django’s admin](https://docs.djangoproject.com/en/3.1/ref/contrib/admin/) interface and a User queryset. This means that using this app, you are able to easily schedule and send emails to the registered users, creating querysets in a friendly interface.  

> For example, you can very quickly configure your app to send a periodic email to the users that haven't logged in in a day or more. Or you can send a custom email to the users that are registered more than a week ago, etc. You only need to define one or more querysets in the admin page.

Reviewer: ```suggestion
For example, you can very quickly configure your app to send a periodic email to the users that haven't logged in in a day or more, send a custom email to the users that registered more than a week ago, etc. To do this, you only need to define one or more querysets in the admin page.
```
I feel a `To do this` or something similar is missing in that next sentence where you explain how to do those examples you mentioned.

If you want to install it and work with it, you can follow the simple steps documented in [the github page of Django Drip Campaigns](https://github.com/rootstrap/django-drip-campaigns).  

> We didn't create this project, we just made a fork of the one written by [Zapier](https://zapier.com/z/qO/) that is no longer maintained. After this we worked on fixing some bugs, doing a couple of enhancements, creating more documentation and publishing the app as a pypi package.  

Reviewer: ```suggestion
We didn't create this project, we just made a fork of the one written by [Zapier](https://zapier.com/z/qO/) that is no longer maintained. After this, we worked on fixing some bugs, doing a couple of enhancements, creating more documentation and publishing the app as a pypi package.  
```

### Features

After a successful installation and configuration, you are able to easily create drip campaigns for email using the admin interface of Django. If you don't have an admin user, create one with the [django createsuperuser command](https://docs.djangoproject.com/en/3.0/intro/tutorial02/#creating-an-admin-user). After this, login into Django's admin page and click on `Drips` to manage them. There you are able to:

- View created and sent drips.
- Create a new drip.
- Select and delete drips.

#### Creating a new drip

Click on the `ADD DRIP +` button to create a new drip. You will see a page like this:

<img src="images/add_drip_page.png" alt="Create Drip" width="700" />

In the next section I will talk about the queryset rules that you can build in an easy way to create very useful drips.

#### Queryset rules

When you create a drip, you need at least one queryset so the users in it will be the ones that receive the defined email. In the `METHOD TYPE` menu you are able to choose to filter or to exclude the users in the defined queryset.  

> On the other hand, when you click in the `FIELD NAME OF THE USER` input, you will see the fields of your User model, and the fields of your user model in the models related to it:

Reviewer: ```suggestion
On the other hand, when you click in the `FIELD NAME OF USER` input, you will see the fields of your User model, and the fields of your user model in the models related to it:
```
Just changing this because the photo shows the column without `THE` word.

Author: `FIELD NAME OF THE USER` is the name in the django admin. I have to keep this name

Reviewer: 👌 As I said, I just commented on this because the photo shows without the `THE` word but no problem leaving it as it is 😄

Author: You are right! my bad

<img src="images/users_lookup_fields.png" alt="Lookup fields" width="350" />

> In the previous image, for example `last_login` is the field in the User model, and `groups__user__id` is the user id from the Groups model that is related with it. So you can enter the name of the field or select it from the list that you see when you click on the input.  

Reviewer: ```suggestion
In the previous image, for example, `last_login` is the field in the User model, and `groups__user__id` is the user id from the Groups model that is related to it. So you can enter the name of the field or select it from the list that you see when you click on the input.  
```

```python
from django.core.mail import EmailMessage
from drip.drips import DripMessage

class PlainDripEmail(DripMessage):

    @property
    def message(self):
        if not self._message:
            email = EmailMessage(self.subject, self.plain, self.from_email, [self.user.email])
            self._message = email
        return self._message
```

In that example, `PlainDripEmail` overrides the message property of the base `DripMessage` class to create a simple `EmailMessage` instance instead of an `EmailMultiAlternatives` instance.

In order to be able to specify that your custom message class should be used for a drip, you need to configure it in the `DRIP_MESSAGE_CLASSES` setting:

```python
DRIP_MESSAGE_CLASSES = {
    'plain': 'myproj.email.PlainDripEmail',
}

```

This will allow you to choose in the admin, for each drip, whether the `default` (`Dripmessage`) or `plain` message class should be used for generating and sending the messages to users.

#### Send drips

To send the created and enabled drips, run the command:

```
python manage.py send_drips
```

You can use cron to schedule the drips.

> ### What we have being doing in the project

Reviewer: ```suggestion
### What we have been doing in the project
```

As I said, we didn't create this project. We forked it and started to make some fixes and improvements. The first thing we did was to upgrade the used Django version. Now it has django 3.0.7.  
Besides we started to create the documentation with [Sphinx](https://www.sphinx-doc.org/en/master/) to have this project in [Read the Docs](https://readthedocs.org/) in the future.  
Another enhancement was to add [flake8](https://flake8.pycqa.org/en/latest/) to enforce the use of style guides in the code. Finally we added an integration with [travis-ci](https://travis-ci.com/) to have a continuous integration service on any change in the project.  

> We registered and fixed some issues such as the view of the User fields, and configuration problems to run the app locally. There are some issues to address and discover yet so if you like OSS your are more than welcome to work with Django Drip Campaigns.  

Reviewer: ```suggestion
We registered and fixed some issues such as the view of the User fields, and configuration problems to run the app locally. There are some issues to address and discover yet so if you like OSS you are more than welcome to work with Django Drip Campaigns.  
```

In the open source initiative we wanted to publish this app as a pypi package, so I'll explain this a little more in the next section.

### Publish a pypi package

> Although there is work to do with the Django Drip Campaigns project, we already published the package in pypi. It's not something very hard to do, but it's important to know a few things before starting, so in this section I'm going to talk about the main points of publishing the app as a package.

Reviewer 2: `Despite Django Drip Campaigns is still a work in progress, we...` in place of `Although...`, sounds more smooth

Reviewer 2: `...the main steps for publishing the app...`

Reviewer 2: Lets try this:

```
Despite Django Drip Campaigns is still a work in progress, we already published the package in PyPI. It's nothing fancy, but it's a must to know a few things before starting, so in this section, I'll be talking about the main steps for publishing an app.
```

Having all set and ready to be published, you need now to [register in the pypi.org page](https://pypi.org/account/register/). After you are registered, go inside the terminal to your project folder and follow these steps:

1. Install `setuptools` and `wheel`. They are used to generate the distribution package: a group of files needed to upload your app.

```
python3 -m pip install --user --upgrade setuptools wheel
```

2. Generate the distribution package. The next command will create all the necessary files under a `dist/` folder:

```
python3 setup.py sdist bdist_wheel
```

3. Install [twine](https://packaging.python.org/key_projects/#twine), that a allows you to upload your app as a package:

```
python3 -m pip install --user --upgrade twine
```

4. Once installed, run twine to upload the archives under `dist/`. This will ask for a user and a password, that are the ones you used on the registration at pypi:

```
python3 -m twine upload --repository pypi dist/*
```

If everything went well, you have published your project as a pypi package. Now you anyone is able to download it with:

```
pip3 install <project-name>
```

### Summary

> In this blog I talked about the Django Drip Campaigns project, the main features and what we did with it at Rootstrap. We love to make contributions to OSS because it's a wonderful way of learning, of sharing knowledge, and creating useful programs to be used by other programmers.  

Reviewer: ```suggestion
In this blog post I talked about the Django Drip Campaigns project, its main features and what we did with it at Rootstrap. We love to make contributions to OSS because it's a wonderful way of learning, of sharing knowledge, and creating useful programs to be used by other programmers.  
```

Besides I talked about the experience and how to publish a project as a pypi package so anyone who wants to give it a try can download it with a simple command.  

> OSS gives a great oportunity not only to see how some other group of people solves a problem addressed by the program, but also to purpose and contribute with fixes and improvements. Finally I hope you have enjoyed this blog and feel motivated to use and work with Django Drip Campaigns and to contribute with OSS.

Reviewer: ```suggestion
OSS gives a great opportunity not only to see how some other group of people solves a problem addressed by the program, but also to purpose and contribute with fixes and improvements. Finally, I hope you have enjoyed this article and feel motivated to use and work with Django Drip Campaigns and to contribute with OSS.
```
